sympy/sets/handlers/functions.py

```diff
@@ -192,7 +192,8 @@ def _set_function(f, self): # noqa:F811
     a = Wild('a', exclude=[n])
     b = Wild('b', exclude=[n])
     match = expr.match(a*n + b)
-    if match and match[a]:
+    if match and match[a] and (not any(
+            i.atoms(Float) for i in (match[a], match[b]))):
```

Comment:
Generally I prefer duplication over the complexity increase of a loop when it's only looping over 2 items:
```python
if match and match[a] and not (match[a].is_Float or match[b].is_Float):
```
It could also be
```python
if match:
    a, b = match[a], match[b]
    # Don't canonicalise with floats
    if not (a.is_Float or b.is_Float):
        ...